### ISSUE TITLE
Angular 5.0.0 error: StaticInjectorError[InjectionToken DocumentToken]

### ISSUE THREAD

**Issue by ebrehault** [angular/angular#20101]
## I'm submitting a...

<pre><code>
[x] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

I have upgraded my current project from 4.4.4 to 5.0.0 using the recommandations from https://angular-update-guide.firebaseapp.com/

And now when I launch `ng serve`, the page keeps blank and the browser console returns the following error:

```
zone.js:391 Uncaught (in promise) Error: StaticInjectorError[InjectionToken DocumentToken]: 
  StaticInjectorError[InjectionToken DocumentToken]: 
    NullInjectorError: No provider for InjectionToken DocumentToken!
```

The traceback indicates it comes from the bootstraping in my `main.ts`:

```
platformBrowserDynamic().bootstrapModule(AppModule)
  .catch(err => console.log(err));
```
but unfortunately, nothing more specific.

## Expected behavior

No error.

## Minimal reproduction of the problem with instructions

I do not know how to reproduce this bug, my existing codebase is quite large.

I am aware it might be difficult to answer my issue, but I just ask to know if this error message rings a bell to anyone, so I know where to start digging.

## Environment

```
Angular CLI: 1.5.0
Node: 8.7.0
OS: darwin x64
Angular: 5.0.0
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

@angular/cdk: 2.0.0-beta.8
@angular/cli: 1.5.0
@angular/flex-layout: 2.0.0-beta.9
@angular/material: 2.0.0-beta.8
@angular-devkit/build-optimizer: 0.0.32
@angular-devkit/core: 0.0.20
@angular-devkit/schematics: 0.0.35
@ngtools/json-schema: 1.1.0
@ngtools/webpack: 1.8.0
@schematics/angular: 0.1.0
typescript: 2.4.2
webpack-bundle-analyzer: 2.9.0
webpack: 3.8.1
```

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [x] Chrome (iOS) version 62.0.3202.62
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: XX  
- Platform:  

Others:

</code></pre>

**Comment by filipbarak** [comment]
I think this error occurs because you need to import HttpModule(deprecated) as well as HttpClientModule, because some dependency still relies on HttpModule.

**Comment by ebrehault** [comment]
Does not seem to make any difference (moreover, if some of the dependencies use the deprecated Http, I guess they import it in their module already). But thanks, though.

**Comment by ochezeau** [comment]
I have the same issue in watch mode.
Everything seems to work properly in production mode.

**Comment by scein** [comment]
I get this too. Any workarounds?

**Comment by Toxicable** [comment]
@filipbarak `Http` is unrelated to this. `HttpClient` does not rely on `Http` nor does either rely on the `DOCUMENT` token.

I tried reproducing the issue but was unable to: https://stackblitz.com/edit/angular-gitter-caoddx?file=app%2Fapp.component.ts

Would someone be able to produce a minimal reproduction that we can investigate the issue with? The issue is likely due to a 3rd party lib , so that's where i'd suggest to start looking, even just searching for use of `DOCUMENT` https://angular.io/api/common/DOCUMENT should point you to where it's being used

**Comment by ebrehault** [comment]
@Toxicable I think I found it, @angular/material version is 2.0.0-beta.8, I guess it is time to upgrade :)

Thank you for your help.

**Comment by Gandii** [comment]
I am getting the same error, checked all my dependencies and none are dependent on HttpModule. Wierd thing is it works if it run it with Cli 1.2.9 with Angular 5 but not with Cli 1.5 and Angular 5.
no difference aot being off or on.

**Comment by sandangel** [comment]
I have the same error. @Gandii how do you use @angular/cli 1.2.9 with angular 5? Angular 5 require @angular/cli 1.5 to compile

**Comment by Gandii** [comment]
@sandangel  was a rather surprised but it worked, did of course upgrade some packages but it was fairly straight forward. 
Didnt try with AOT though.

**Comment by ochezeau** [comment]
I fixed my problem.
For me, everything was working well with a : build --prod
I had the problem using a common library (unsing an npm link) and with the command ng build --watch (I use a tomcat server)
I fixed the problem using the command: npm build --aot --watch (adding --aot)
Is is alos important to have the right version of components : angular-cli : 1.5 and typescript 2.4.2

**Comment by trotyl** [comment]
@ochezeau You'll need `--preserve-symlinks` when link to a different repo. That's what happened in #20102.

**Comment by ochezeau** [comment]
@trotyl : i removed my flag --aot and i replace it by --preserve-symlinks and it works fine. Thank you

**Comment by vicb** [comment]
5.1 will gives more context for injection errors.

**Comment by ThalesMatoso** [comment]
Same error, but only works on some components and routes, any solution?

**Comment by zuo305** [comment]
import {HttpClientModule, HttpClient} from '@angular/common/http';

then:
imports:[
HttpClientModule]

will be ok

**Comment by Toxicable** [comment]
If you have found a bug then opening a new issue for that bug would be the best action here.
Having a minimal reproduction showing Angular at fault is an important thing to have

**Comment by casvil** [comment]
Had the same issue when creating a custom directive using and importing Window without declaring it in providers and trying to access it inside a component

**Comment by usmanakram1996** [comment]
Add this in app.module.ts 

**import { HttpClientModule } from '@angular/common/http'**

and also add HttpClientModule in your **imports**

**Comment by StasGoshtein** [comment]
Everything might be much simpler - for me, the solution was to fix my code. "ngIf" without leading asterisk will cause this exception to be thrown.

**Comment by pragadeesh2050** [comment]
Yes adding these below to app.module.ts worked. Thanks @usmanakram1996 

import { HttpModule } from '@angular/http';

and importing worked


imports: [
     HttpModule
  ]

**Comment by Sagar1990** [comment]
@pragadeesh2050 , can you please share your system.config.js.
I am facing the same issue and seems like i am missing something over there.

**Comment by Sagar1990** [comment]
I am getting the same error, including Httpmodule did not work for me.

my app.module.ts: -
```

import { NgModule, Injectable } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { FormsModule } from '@angular/forms';
import { RouterModule, Routes } from '@angular/router';
import { HttpModule } from '@angular/http';


import { ModalModule } from 'ngx-modialog';
import { BootstrapModalModule, Modal, bootstrap4Mode } from 'ngx-modialog/plugins/bootstrap';

import { AppComponent } from './app.component';
import { ExportToPdfComponent } from './exportTopdf/exportTopdf.component';
import { InvalidPageComponent } from './invalidPage/invalidPage.component';

import { ViewCountService } from './Service/viewsCount.component';

const appRoutes: Routes = [
    { path: 'home', component: AppComponent },
    { path: '', redirectTo: '/home', pathMatch: 'full' },
    { path: 'export', component: ExportToPdfComponent },
    { path: '**', component: InvalidPageComponent }
];

@NgModule({
    imports: [
        BrowserModule,
        HttpModule,
        RouterModule.forRoot(appRoutes),
        ModalModule.forRoot(),
        BootstrapModalModule
    ],
    declarations: [
        AppComponent,
        ExportToPdfComponent,
        InvalidPageComponent
    ],
    bootstrap: [
        AppComponent
    ],
    providers: [
        ViewCountService
    ]
})



export class AppModule { }

```

exportTopdf.component.ts file: -

```
import { Component, ViewContainerRef, ViewEncapsulation } from '@angular/core';
import { ViewCountService } from '../Service/ViewsCount.component';

import { Overlay } from 'ngx-modialog';
import { Modal } from 'ngx-modialog/plugins/bootstrap';

@Component({
    selector: 'export-to-pdf',
    templateUrl: 'app/exportTopdf/exportTopdf.component.html',
})
export class ExportToPdfComponent {
    name: string;
    fields: any;

    constructor(public modal: Modal, private ViewCountService: ViewCountService) {
        debugger;
        this.fields = [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }];
    }


    getViewCount(): void {
        debugger;
        this.ViewCountService.getViewCount()
            .then(data => {
                this.name = data;
                console.log("I CANT SEE DATA HERE: ", this.name)
            });
    }

}

```

my service code: -
```

import { Injectable } from '@angular/core';
import { Http } from '@angular/http';
import { Observable } from 'rxjs/Rx';

@Injectable()
export class ViewCountService {
    constructor(private http: Http) {
    }

    getViewCount() {
        return this.http.get('api/Tableau/GetViewsCount')
            .map(response => response.json() as string).toPromise();
    }

    getDataObservable(url: string) {
        return this.http.get('api/Tableau/GetViewsCount')
            .map(data => {
                data.json();
                console.log("I CAN SEE DATA HERE: ", data.json());
            });
    }

}
```

system.config.js: -
```

/**
 * System configuration for Angular samples
 * Adjust as necessary for your application needs.
 */
(function (global) {
    System.config({
        transpiler: 'typescript',
        //typescript compiler options
        typescriptOptions: {
            emitDecoratorMetadata: true
        },
        paths: {
            // paths serve as alias
            'npm:': '/node_modules/'
        },
        // map tells the System loader where to look for things
        map: {
            // our app is within the app folder
            'app': 'app',

            // angular bundles
            '@angular/core': 'npm:@angular/core/bundles/core.umd.js',
            '@angular/common': 'npm:@angular/common/bundles/common.umd.js',
            '@angular/compiler': 'npm:@angular/compiler/bundles/compiler.umd.js',
            '@angular/platform-browser': 'npm:@angular/platform-browser/bundles/platform-browser.umd.js',
            '@angular/platform-browser-dynamic': 'npm:@angular/platform-browser-dynamic/bundles/platform-browser-dynamic.umd.js',
            '@angular/http': 'npm:@angular/http/bundles/http.umd.js',
            '@angular/router': 'npm:@angular/router/bundles/router.umd.js',
            '@angular/forms': 'npm:@angular/forms/bundles/forms.umd.js',

            // other libraries
            'rxjs': 'npm:rxjs',
            'angular-in-memory-web-api': 'npm:angular-in-memory-web-api/bundles/in-memory-web-api.umd.js',

            'ngx-modialog': 'npm:ngx-modialog/bundle/ngx-modialog.umd.min.js',
            'ngx-modialog/plugins/bootstrap': 'npm:ngx-modialog/plugins/bootstrap/bundle/ngx-modialog-bootstrap.umd.min.js'


        },
        // packages tells the System loader how to load when no filename and/or no extension
        packages: {
            app: {
                defaultExtension: 'js',
                meta: {
                    './*.js': {
                        loader: 'systemjs-angular-loader.js'
                    }
                }
            },
            rxjs: {
                defaultExtension: 'js'
            }
        }
    });
})(this);

```

package.json: -

```
{
  "name": "angular-quickstart",
  "version": "1.0.0",
  "description": "QuickStart package.json from the documentation, supplemented with testing support",
  "scripts": {
    "build": "tsc -p src/",
    "build:watch": "tsc -p src/ -w",
    "build:e2e": "tsc -p e2e/",
    "serve": "lite-server -c=bs-config.json",
    "serve:e2e": "lite-server -c=bs-config.e2e.json",
    "prestart": "npm run build",
    "start": "concurrently \"npm run build:watch\" \"npm run serve\"",
    "pree2e": "npm run build:e2e",
    "e2e": "concurrently \"npm run serve:e2e\" \"npm run protractor\" --kill-others --success first",
    "preprotractor": "webdriver-manager update",
    "protractor": "protractor protractor.config.js",
    "pretest": "npm run build",
    "test": "concurrently \"npm run build:watch\" \"karma start karma.conf.js\"",
    "pretest:once": "npm run build",
    "test:once": "karma start karma.conf.js --single-run",
    "lint": "tslint ./src/**/*.ts -t verbose"
  },
  "keywords": [],
  "author": "",
  "license": "MIT",
  "dependencies": {
    "@angular/common": "~5.0.3",
    "@angular/compiler": "~5.0.3",
    "@angular/core": "~5.0.3",
    "@angular/forms": "~5.0.3",
    "@angular/http": "~5.0.3",
    "@angular/platform-browser": "~5.0.3",
    "@angular/platform-browser-dynamic": "~5.0.3",
    "@angular/router": "~5.0.3",
    "angular-in-memory-web-api": "~0.5.1",
    "bootstrap": "^3.3.7",
    "core-js": "^2.4.1",
    "ngx-modialog": "^5.0.0",
    "rxjs": "5.5.2",
    "systemjs": "0.20.19",
    "zone.js": "^0.8.4"
  },
  "devDependencies": {
    "concurrently": "^3.2.0",
    "lite-server": "^2.2.2",
    "typescript": "~2.6.1",
    "canonical-path": "0.0.2",
    "tslint": "^5.8.0",
    "lodash": "^4.16.4",
    "jasmine-core": "~2.8.0",
    "karma": "^1.3.0",
    "karma-chrome-launcher": "^2.0.0",
    "karma-cli": "^1.0.1",
    "karma-jasmine": "^1.0.2",
    "karma-jasmine-html-reporter": "^0.2.2",
    "protractor": "~5.2.0",
    "rimraf": "^2.5.4",
    "@types/node": "^8.0.53",
    "@types/jasmine": "2.8.2"
  },
  "repository": {}
}

```


Error i am getting is: -
```

ERROR Error: Uncaught (in promise): Error: StaticInjectorError[ViewCountService]: 
  StaticInjectorError[ViewCountService]: 
    NullInjectorError: No provider for ViewCountService!
Error: StaticInjectorError[ViewCountService]: 
  StaticInjectorError[ViewCountService]: 
    NullInjectorError: No provider for ViewCountService!
    at _NullInjector.get (injector.js:31)
    at resolveToken (injector.js:387)
    at tryResolveToken (injector.js:330)
    at StaticInjector.get (injector.js:170)
    at resolveToken (injector.js:387)
    at tryResolveToken (injector.js:330)
    at StaticInjector.get (injector.js:170)
    at resolveNgModuleDep (ng_module.js:103)
    at NgModuleRef_.get (refs.js:1037)
    at resolveDep (provider.js:455)
    at _NullInjector.get (injector.js:31)
    at resolveToken (injector.js:387)
    at tryResolveToken (injector.js:330)
    at StaticInjector.get (injector.js:170)
    at resolveToken (injector.js:387)
    at tryResolveToken (injector.js:330)
    at StaticInjector.get (injector.js:170)
    at resolveNgModuleDep (ng_module.js:103)
    at NgModuleRef_.get (refs.js:1037)
    at resolveDep (provider.js:455)
    at resolvePromise (zone.js:824)
    at resolvePromise (zone.js:795)
    at zone.js:873
    at ZoneDelegate.invokeTask (zone.js:425)
    at Object.onInvokeTask (ng_zone.js:575)
    at ZoneDelegate.invokeTask (zone.js:424)
    at Zone.runTask (zone.js:192)
    at drainMicroTaskQueue (zone.js:602)
    at ZoneTask.invokeTask [as invoke] (zone.js:503)
    at invokeTask (zone.js:1540)
defaultErrorLogger @ errors.js:48
ErrorHandler.handleError @ error_handler.js:90
next @ application_ref.js:311
schedulerFn @ event_emitter.js:156
SafeSubscriber.__tryOrUnsub @ Subscriber.ts:254
SafeSubscriber.next @ Subscriber.ts:204
Subscriber._next @ Subscriber.ts:135
Subscriber.next @ Subscriber.ts:95
Subject.next @ Subject.ts:61
EventEmitter.emit @ event_emitter.js:131
(anonymous) @ ng_zone.js:605
ZoneDelegate.invoke @ zone.js:392
Zone.run @ zone.js:142
NgZone.runOutsideAngular @ ng_zone.js:404
onHandleError @ ng_zone.js:605
ZoneDelegate.handleError @ zone.js:396
Zone.runGuarded @ zone.js:158
_loop_1 @ zone.js:702
api.microtaskDrainDone @ zone.js:711
drainMicroTaskQueue @ zone.js:610
ZoneTask.invokeTask @ zone.js:503
invokeTask @ zone.js:1540
globalZoneAwareCallback @ zone.js:1566
```

**Comment by usmanakram1996** [comment]
@Sagar1990 have you imported this 

**import { HttpClientModule } from '@angular/common/http'**

in app.module.ts?

**Comment by mani9818** [comment]
No need to import **HttpModule**
**import {HttpModule} from '@angular/http'**   -- Not Required

Just make sure, you have provided your service in providers after importing it in app.module.ts
Like - 

import { ApiService } from './app.service'
imports: [
    BrowserModule,
  ],
providers: [ApiService],

**Comment by nayfin** [comment]
I got this after importing a component library I built with ng-packagr. I had to change my dependencies to peer dependencies in my component library as outlined [here](https://blog.angularindepth.com/making-your-angular-2-library-statically-analyzable-for-aot-e1c6f3ebedd5#7921). This cost me days to root out.

**Comment by jvgpinto** [comment]
Thanks @maniwadhwa26 it works for me...

**Comment by MPauletta** [comment]
Thanks @maniwadhwa26 it works for me too. :-)

**Comment by sal0w** [comment]
Thanks, @maniwadhwa26  Works for me tooo 👍

**Comment by en-sevilla** [comment]
I got the error after updating from angularfire2  (4 to 5). I solved by replacing the following in app.module.ts

`import { AngularFireDatabaseModule } from 'angularfire2/database';`

by

`import { AngularFireDatabaseModule } from 'angularfire2/database-deprecated'`

My current package.json config (I last more than 4 hours to update from 4 to 5 and other several major updating, so this conf included some downgrading corrections -like typescript-)

```
{
  "name": "FirebaseCRUD",
  "version": "0.0.1",
  "author": "Ionic Framework",
  "homepage": "http://ionicframework.com/",
  "private": true,
  "scripts": {
    "clean": "ionic-app-scripts clean",
    "build": "ionic-app-scripts build",
    "lint": "ionic-app-scripts lint",
    "ionic:build": "ionic-app-scripts build",
    "ionic:serve": "ionic-app-scripts serve"
  },
  "dependencies": {
    "@angular/common": "5.1.3",
    "@angular/compiler": "5.1.3",
    "@angular/compiler-cli": "5.1.3",
    "@angular/core": "5.1.3",
    "@angular/forms": "5.1.3",
    "@angular/http": "5.1.3",
    "@angular/platform-browser": "5.1.3",
    "@angular/platform-browser-dynamic": "5.1.3",
    "@angular/tsc-wrapped": "^4.2.6",
    "@ionic-native/app-minimize": "^4.3.2",
    "@ionic-native/autostart": "^4.5.2",
    "@ionic-native/core": "4.5.2",
    "@ionic-native/date-picker": "^4.4.2",
    "@ionic-native/geolocation": "^4.3.3",
    "@ionic-native/native-storage": "^4.4.0",
    "@ionic-native/network": "^4.5.2",
    "@ionic-native/splash-screen": "4.5.2",
    "@ionic-native/status-bar": "4.5.2",
    "@ionic/storage": "2.1.3",
    "angularfire2": "^5.0.0-rc.5-next",
    "cordova-android": "7.0.0",
    "cordova-browser": "~5.0.1",
    "cordova-plugin-appminimize": "^1.0.0",
    "cordova-plugin-autostart": "^2.0.1",
    "cordova-plugin-compat": "^1.2.0",
    "cordova-plugin-console": "^1.0.5",
    "cordova-plugin-device": "2.0.1",
    "cordova-plugin-geolocation": "4.0.1",
    "cordova-plugin-nativestorage": "^2.2.2",
    "cordova-plugin-network-information": "2.0.1",
    "cordova-plugin-splashscreen": "5.0.1",
    "cordova-plugin-statusbar": "^2.2.2",
    "cordova-plugin-whitelist": "^1.3.1",
    "firebase": "4.8.0",
    "ionic-angular": "3.9.2",
    "ionic-plugin-keyboard": "^2.2.1",
    "ionic2-calendar": "^0.3.13",
    "ionicons": "3.0.0",
    "latest": "^0.2.0",
    "moment": "^2.20.0",
    "rxjs": "^5.5.6",
    "sw-toolbox": "3.6.0",
    "typescript": "2.6.2",
    "zone.js": "0.8.19"
  },
  "devDependencies": {
    "@ionic/app-scripts": "3.1.7",
    "@ionic/cli-plugin-ionic-angular": "1.4.1",
    "typescript": "2.6.2"
  },
  "description": "An Ionic project",
  "cordova": {
    "plugins": {
      "cordova-plugin-console": {},
      "cordova-plugin-device": {},
      "cordova-plugin-splashscreen": {},
      "cordova-plugin-statusbar": {},
      "cordova-plugin-whitelist": {},
      "ionic-plugin-keyboard": {},
      "cordova-plugin-appminimize": {},
      "cordova-plugin-geolocation": {
        "GEOLOCATION_USAGE_DESCRIPTION": "Control de Presencia"
      },
      "cordova-plugin-nativestorage": {},
      "cordova-plugin-network-information": {},
      "cordova-plugin-autostart": {}
    },
    "platforms": [
      "android",
      "browser"
    ]
  }
}

```

**Comment by gangachris** [comment]
I had the same error, but my issue turned out to be using a wrong format for a route resolver. I had an array instead of an object.

Hope this helps.

**Comment by kverma17** [comment]
Not working with any of the solutions above

Attached are the files---
![appmodulets](https://user-images.githubusercontent.com/35423116/35138647-956ef212-fd15-11e7-8d0e-f8e04b2a7a6b.png)
![homets](https://user-images.githubusercontent.com/35423116/35138648-959f0092-fd15-11e7-864d-54e1dc62a58f.png)
![packagejson](https://user-images.githubusercontent.com/35423116/35138649-95e52c48-fd15-11e7-9240-d8c02cacd430.png)

**Comment by nayfin** [comment]
@kverma17 You only need to change your `dependencies` to `peerDependencies` if you are building a library. You might try updating all angular dependencies to latest, I know better error definitions for this error were a part of 5.2 or 5.1. I am guessing some library you are dependent on changed their syntax without documenting it as happened to en-sevilla. Firebase hase been bad about making these kinds of changes lately, I see you importing firebase in your app module but I don't see it in your `package.json`. If you are using `"firebase": "4.8.1"` roll it back to 4.8.0 as they made breaking changes to the way the modules are imported in 4.8.1.

**Comment by izverev-aexsoft** [comment]
As for me, it started to happen once I accidentially imported CommonModule instead of BrowserModule in my main app module. I was in the middle of mass refactoring and it's very easy to fail, considering that 1) you may include anything anywhere, 2) nothing is required, 3) once you did something wrong, errors you get usually are too obscure... Hope it'll get better soon.

**Comment by imnickvaughn** [comment]
If the above solutions haven't worked for you, try this. I hope a better solution comes along but for now

**Comment by imnickvaughn** [comment]
Are you guys importing AngularFirebaseDatabase?  Did this break old code?  I can not see some of your guys' code but if the above solutions have not worked for you give this a shot and tell us if it works (waiting for a better solution but this will get you going again at least).

All you have to do is replace your AngularFirebaseDatabase imports with this...  

`import {AngularFireDatabase} from 'angularfire2/database-deprecated';`

so essentially you are adding '-deprecated'

hope this helps

**Comment by NurGuz23** [comment]
Hi,

I don't know how to add  --preserve-symlinks

**Comment by thatsrohitnaik** [comment]
Helllo.. there... 
hmm with regards to this error if any of your are using services then please make sure that it has been imported in the app.module.ts

**Comment by Wordeur** [comment]
I have the same problem with Angular 5.2.7
When i try to add : 

```
 import {ActivatedRoute} from "@angular/router";

 constructor(private router: ActivatedRoute) {}
```

In my component, this generates an error : 

> AppComponent_Host.ngfactory.js? [sm]:1 ERROR Error: StaticInjectorError(AppModule)[AppComponent -> ActivatedRoute]: 
>   StaticInjectorError(Platform: core)[AppComponent -> ActivatedRoute]: 
>     NullInjectorError: No provider for ActivatedRoute!

I have try all solution exposed in above but no works...

**Comment by mlc-mlapis** [comment]
@Wordeur ... it certainly works ... so you have to have a problem in your code which you simply don't see ... so prepare a simple Stackblitz reproduction demo ... and you will see that it works.

**Comment by shaharido1** [comment]
Hey, I'm struggling with similar issue as this stackOverFlow question - [Angular can't use HttpClient in external library](https://stackoverflow.com/questions/48174555/angular-cant-use-httpclient-in-external-library). In short - i'm trying to compile my app as a ng-module (using ngc), so it could be imported by another outer app. Unfortunately, though my app works fine by itself; it can't be imported by the outer app, due to the error mentioned above. None of the solution suggested above solved the problem. Help would be appreciated.

**Comment by xwa130** [comment]
I solve this by following two steps: 
> import { BrowserModule }  from '@angular/platform-browser';
...
  imports: [
    BrowserModule,
  ]

**Comment by kambleaa007** [comment]
Stuck here The hell
Fuk this bitch .......
Help out guys with exact matching solution all solutions are hot as hell..............
NEED PROPER HELP
ERROR Error: StaticInjectorError(AppModule)[AppComponent -> AngularFireDatabase]: 
  StaticInjectorError(Platform: core)[AppComponent -> AngularFireDatabase]: 
    NullInjectorError: No provider for AngularFireDatabase!
    at _NullInjector.get (core.js:1002)
    at resolveToken (core.js:1300)
    at tryResolveToken (core.js:1242)
    at StaticInjector.get (core.js:1110)
    at resolveToken (core.js:1300)
    at tryResolveToken (core.js:1242)
    at StaticInjector.get (core.js:1110)
    at resolveNgModuleDep (core.js:10854)
    at NgModuleRef_.get (core.js:12087)
    at resolveDep (core.js:12577)

**Comment by amit-kumar27** [comment]
@Toxicable 
I am trying to upgrade angular version from angular 2.0 to angular 5.0 and now I am stuck with this same error. Tried some of the solution provided here, but still not able to fix it.
This is the sample app with the same architecture we are following for our application:-
https://github.com/amit-kumar27/sample-app
Please take a look and provide a solution.

**Comment by madhavsharmaagra** [comment]
All you need to do is to import your service in your app.module, like,

__import { HeroService } from '../app/hero.service';__

and then, provide it to your providers array in the same file, like,

__providers: [HeroService]__

I hope this will help....

**Comment by amit-kumar27** [comment]
@madhavsharmaagra Thanks for the help, but if you see the sample-app, there is no service in my sample-app. So it is not applicable.

**Comment by sekurukaguvi** [comment]
removing the following from my provider page helped,

**import { HttpClient } from '@angular/common/http';**
constructor(**public http: HttpClient**) {}

**Comment by asrar7** [comment]
@filipbarak @Toxicable @ebrehault @ochezeau @sekurukaguvi 
Hi guys, I have somehow the same problem after importing both Http and HttpClient. Please help me . thanks 
![image](https://user-images.githubusercontent.com/39076095/40153099-ae068522-59ba-11e8-9c08-c410d380f5fa.png)

**Comment by infact** [comment]
I already had  these import  statements inside my app.module.But still the StaticInjectorError ...NullInjectorError poped up.

import { BrowserModule } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { HttpClientModule } from '@angular/common/http';

**THIS IS HOW I GOT IT RESOLVED**
The error happened because I forgot to import my service into app.module.ts file.

`import {xyzNotificationService } from './services/eventoccurrences.service';`

and then added _xyzNotificationService_ to the _providers: [ ]_ section of the app.module

**Comment by ledicjp** [comment]
@nayfin how did you manage to package it with ng-packagr if you have empty dependencies?
If I rename dependencies to peerDependencies my ng-packagr doesn't find @angular/core etc. anymore

**Comment by nayfin** [comment]
@ledicjp are you talking about when you build or when you package. At the moment I change the parameter name to `dependencies` when building for development and then to `peerDependencies` when I am packaging for release. 

I started this library before ng 6 implemented its own library helper utility and I still have a lot to learn in this area. I plan on rebuilding soon and digging into this more. I'll update post then. Hope this at least gets you unblocked.

**Comment by nayfin** [comment]
@ledicjp I am coming to realize that I am just going to have to rebuild using the cli. [Here](https://blog.angularindepth.com/creating-a-library-in-angular-6-87799552e7e5) is a walk-through for anyone else who's coming to the same realization. Hopefully not too painful.

**Comment by AhmedYousseff** [comment]
this error could be indication that you didn't register your service as a providers in the app.module.ts

**Comment by TheSegfault** [comment]
Is this a troll post XD

**Comment by serhioV** [comment]
> import {HttpClientModule, HttpClient} from '@angular/common/http';
> 
> then:
> imports:[
> HttpClientModule]
> 
> will be ok

Where to import this? in the main module or the component one?

**Comment by nayfin** [comment]
@ugurkoysuren i believe the ng g library functionality was added in v6.

**Comment by sorcamarian** [comment]
In my case, a `Route guard` was not added to the `providers: any[]` inside of `@ngModule`;

```
@NgModule({
  declarations: [
    BookingFormComponent,
    AppComponent,
    HomeComponent,
    BookingNotFoundComponent
  ],
  imports: [
    BrowserModule,
    RouterModule.forRoot(appRoutes),
    ReactiveFormsModule
  ],
  providers: [BookingFormGuard],
  bootstrap: [AppComponent]
})
export class AppModule { }
```
